core/cas-server-core-tickets-api/src/main/java/org/apereo/cas/ticket/factory/DefaultProxyGrantingTicketFactory.java

```diff
@@ -75,6 +86,31 @@
         return (T) result;
     }
 
+    /**
+     * Produce the ticket with the adequate expiration policy.
+     *
+     * @param servicePgtPolicy the proxy granting ticket expiration policy
+     * @param serviceTicket    the service ticket
+     * @param authentication   the authentication
+     * @param pgtId            the PGT id
+     * @return the ticket
+     */
+    protected ProxyGrantingTicket produceTicketWithAdequateExpirationPolicy(
+            final RegisteredServiceProxyGrantingTicketExpirationPolicy servicePgtPolicy,
+            final ServiceTicket serviceTicket,
+            final Authentication authentication,
+            final String pgtId) {
+        if (servicePgtPolicy != null) {
+            LOGGER.trace("Overriding proxy-granting ticket policy with the specific policy: [{}]", servicePgtPolicy);
+            return serviceTicket.grantProxyGrantingTicket(pgtId, authentication,
+                    new HardTimeoutExpirationPolicy(servicePgtPolicy.getMaxTimeToLiveInSeconds()));
```

Comment:
This looks perfectly fine. Some future refactoring may be in order if we could use a specific builder class and not duplicate the logic here with a hard-coded type of timeout. In theory, because TGT policies can be the same as PGTs by default, we should try to allow for the same type of options when we define policies on a per-service basis as well. (i.e. if a policy can be based on a idle-timeout for TGT per service, it should also be allowed for the same thing when defined for a PGT per service)

Reply:
As you say "future", I understand that this pull request remains "as is" for the merge, right?

This is something I thought about, but to be honest I wan't sure about the expected implementation. As the `buildTicketExpirationPolicy` method does not accept any parameter, I couldn't think of any solution except than creating **every time** an expiration policy builder from the service PGT policy. Something like:

```java
if (servicePgtPolicy != null) {
  LOGGER.trace("Overriding proxy-granting ticket policy with the specific policy: [{}]", servicePgtPolicy);
  val servicePGTExpirationPolicyBuilder =
new RegisteredServicePGTExpirationPolicyBuilder(servicePgtPolicy);
  return serviceTicket.grantProxyGrantingTicket(pgtId, authentication, 
servicePGTExpirationPolicyBuilder.buildTicketExpirationPolicy());
```

Is this what you had in mind?

Reply:
> As you say "future", I understand that this pull request remains "as is" for the merge, right?

Exactly, yes. Let's not worry about this right now. 

> Is this what you had in mind?

Yes, but after your explanation it's obvious that might need to clean up the API slightly to allow for dynamic building of expiration policies without having to create the builder every time. Something to do later perhaps once the need pops up, and we can make a note of it internally to revisit perhaps in RC4 or 6.3, etc.

Thanks for your review!